### PR TITLE
refactor: Make published artifact names consistent and use permalinks

### PR DESCRIPTION
**Pull request by ReactorScram**
```[tasklist]
- [x] Update website
- [x] Update blog entry with old link
- [ ] ~~Replace Github URL in GUI Client updater with our own links~~
- [ ] Wait for CI to go green
```

Refs #4531 

This proposes a unified scheme for deb and MSI packages, and moves Windows to that scheme.

This breaks compatibility. Existing Clients won't recognize the new asset names once this is merged, so they won't show the "Firezone 1.0.0 is available" pop-up.

### CODE REVIEW

File: .github/workflows/_tauri.yml
Comment (ReactorScram) [inline .github/workflows/_tauri.yml:39]: Putting the OS name here is redundant since every package type (MSI, dmg, deb) belongs to exactly one OS, but I think it's easier to read "firezone-client-gui-windows" and stop there than to read "firezone-client-gui.......... msi"

Comment (jamilbk) [reply .github/workflows/_tauri.yml:39]: Makes sense

**Review by jamilbk — APPROVED** [review]
LGTM -- let's wait to merge until I have the website stuff in.

Should I open a new PR or would you prefer I build off this one?

File: .github/workflows/_tauri.yml
Comment (jamilbk) [inline .github/workflows/_tauri.yml:42]: If we're changing these to `x86_64` we should probably do the gateway too right?

Comment (ReactorScram) [reply .github/workflows/_tauri.yml:42]: Yeah. It feels a bit pointless to change to "x86_64" at all because I could definitely map "x86_64" to "x64" in the Rust code. I'm just doing it so that the update code can use Rust's `std::env::consts::ARCH` directly. But maybe it's not worth it?

**Comment by ReactorScram** [comment]
@jamilbk You can add the website commits to this PR, there shouldn't be any conflicts

File: .github/workflows/_tauri.yml
Comment (jamilbk) [reply .github/workflows/_tauri.yml:42]: Let's change it then -- we'll soon have permalinks anyway

Comment (jamilbk) [reply .github/workflows/_tauri.yml:42]: Well, hmmm that'll require changing a bit of CI configuration. I'll see if I can do it in my commits.

File: .github/workflows/_build_artifacts.yml
Comment (jamilbk) [inline .github/workflows/_build_artifacts.yml]: @ReactorScram Tried to match the new Windows package name -- did I succeed?

File: .github/workflows/_build_artifacts.yml
Comment (jamilbk) [inline .github/workflows/_build_artifacts.yml]: Other component packages are named here.

File: scripts/build/tauri-upload-windows.sh
Comment (jamilbk) [inline scripts/build/tauri-upload-windows.sh:8]: Just verifying this will be e.g. `firezone-client-windows-gui_1.0.0_x86_64.msi`

File: scripts/gateway-systemd-install.sh
Comment (jamilbk) [inline scripts/gateway-systemd-install.sh]: See #4753

File: website/redirects.js
Comment (jamilbk) [inline website/redirects.js:1]: Added both `latest` and versioned permalinks here.

File: rust/gui-client/src-tauri/src/client/updates.rs
Comment (jamilbk) [inline rust/gui-client/src-tauri/src/client/updates.rs:95]: We could avoid hitting the GitHub API now, using the following algorithm:

1. Check what `https://www.firezone.dev/dl/firezone-client-windows-gui/latest/x86_64` redirects to
2. Parse version from redirected URL
3. If a new version is available, alert the user to download using this URL

**Review by jamilbk — COMMENTED** [review]
@ReactorScram This is ready for your review. I'll do a bit more testing, but this gives us permalinks that we can use for the future.

They have the format

```
https://www.firezone.dev/dl/:package/:version/:arch
```

We can add to them in the future, such as the `:package_type` (i.e. deb or exe)

**Comment by jamilbk** [comment]
@ReactorScram I purposefully did not use `1.0.0-pre.14` here and instead `1.0.0` because Cargo doesn't allow the former.

Since we need to time this with the next release anyway, I'm going full steam ahead towards actual 1.0.0 since we have portal updates to deploy too.

File: .github/workflows/_build_artifacts.yml
Comment (ReactorScram) [reply .github/workflows/_build_artifacts.yml]: It should be "firezone-client-headless-linux" I think, since "client-headless" is the product or package

**Comment by ReactorScram** [comment]
@jamilbk But we've had all those `-pre` releases on Github, those would all alias to 1.0.0 under this system?

**Comment by jamilbk** [comment]
> @jamilbk But we've had all those `-pre` releases on Github, those would all alias to 1.0.0 under this system?

Those won't work anyway because the package name is changing too right? The old download links will still work because we aren't editing past releases, so I'm not sure we need to burden this new system with the `pre` packages.

File: .github/workflows/_build_artifacts.yml
Comment (jamilbk) [reply .github/workflows/_build_artifacts.yml]: Hm, this is confusing though:

- `firezone-client-windows-gui`
- `firezone-client-headless-linux`

The platform and operating mode are swapped. That feels a bit like our cargo directory structure is leaking through to the user.

Comment (jamilbk) [reply .github/workflows/_build_artifacts.yml]: Or am I misunderstanding?

**Comment by ReactorScram** [comment]
@jamilbk Okay. So there won't be anything like 1.0.1-pre.1 and 1.0.1-pre.2?

File: .github/workflows/_build_artifacts.yml
Comment (jamilbk) [reply .github/workflows/_build_artifacts.yml]: Ah yeah, I think I got it turned around. More coffee time

**Comment by jamilbk** [comment]
> @jamilbk Okay. So there won't be anything like 1.0.1-pre.1 and 1.0.1-pre.2?

Oh gosh I certainly hope not. Nah the plan is to more eagerly bump versions going forward, and then at some point when we having some breathing room decouple the publishing steps for them and have releases like `https://github.com/firezone/firezone/releases/download/firezone-gateway_1.0.0/firezone-gateway_1.0.0_x86_64` where we perform individual releases.

Like https://github.com/tokio-rs/tracing-appender does it

File: .github/workflows/_tauri.yml
Comment (jamilbk) [reply .github/workflows/_tauri.yml:42]: Updated

File: .github/workflows/_build_artifacts.yml
Comment (jamilbk) [reply .github/workflows/_build_artifacts.yml]: Fixed

File: scripts/build/tauri-upload-windows.sh
Comment (jamilbk) [reply scripts/build/tauri-upload-windows.sh:8]: Edit, `firezone-client-gui-windows_1.0.0_x86_64.msi`

File: rust/gui-client/src-tauri/src/client/updates.rs
Comment (ReactorScram) [reply rust/gui-client/src-tauri/src/client/updates.rs:95]: ~~Should I try to do that change in this same PR?~~

I'll try adding that in here

File: scripts/gateway-systemd-install.sh
Comment (ReactorScram) [inline scripts/gateway-systemd-install.sh:62]: ~~I'll replace the Github URL in the GUI Client update check with this~~

File: rust/gui-client/src-tauri/src/client/updates.rs
Comment (jamilbk) [reply rust/gui-client/src-tauri/src/client/updates.rs:95]: It may be a bit tricky to test. If you want to test locally you can try running the website locally:

```
cd website/
pnpm i
pnpm dev
```

Comment (ReactorScram) [reply rust/gui-client/src-tauri/src/client/updates.rs:95]: Yeah it got complicated. I'll make it a separate PR that stacks on this.

**Comment by ReactorScram** [comment]
@jamilbk Looks good. I changed "armv7l" to "armv7" internally to make it consistent with Rust's arch naming.
Hopefully we didn't miss anything.

This must also be waiting for CI to go green, right?

**Comment by jamilbk** [comment]
> This must also be waiting for CI to go green, right?

After this gets merged to `main` we'll want to immediately publish the `1.0.0` release

**Comment by ReactorScram** [comment]
Okay so I don't want to merge it right now?

**Comment by jamilbk** [comment]
> Okay so I don't want to merge it right now?

Not yet -- doing so would publish the website right away and update the gateway upgrade script which will contain broken links since the release isn't live yet. Gotta do it the other way around.

**Comment by jamilbk** [comment]
Ah I goofed, we need to split this up so that the website points to the current naming scheme until the release contains the updated asset names.

**Comment by jamilbk** [comment]
Nevermind, looks like we can simply change the asset names on the published release 😎

**Pull request merged**
